security/pkg/server/ca/authenticate/xfcc_authenticator.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package authenticate
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/peer"
+
+	"github.com/alecholmes/xfccparser"
+
+	"istio.io/istio/pkg/security"
+)
+
+const (
+	XfccAuthenticatorType = "XfccAuthenticator"
+)
+
+// XfccAuthenticator extracts identities from Xfcc header.
+type XfccAuthenticator struct{}
+
+var _ security.Authenticator = &XfccAuthenticator{}
+
+func (xff *XfccAuthenticator) AuthenticatorType() string {
+	return XfccAuthenticatorType
+}
+
+// Authenticate extracts identities from Xfcc Header.
+func (xff *XfccAuthenticator) Authenticate(ctx security.AuthContext) (*security.Caller, error) {
+	// TODO: Extend this for other "trusted" IPs based on network policies.
+	if !isLocalHost(ctx.RequestContext) {
+		return nil, fmt.Errorf("call is not from trusted network, xfcc can not be used as authenticator")
```

Comment:
I really don't like the fact we treat normal occurance as error. Now that we call all auth ticators we can remove the error - there is nothing wrong about not having xfcc header.